src/data/polynomial/lifts.lean

```diff
@@ -0,0 +1,315 @@
+/-
+Copyright (c) 2020 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Riccardo Brasca
+-/
+
+import data.polynomial.algebra_map
+import algebra.algebra.subalgebra
+import algebra.polynomial.big_operators
+import data.polynomial.erase_lead
+import data.polynomial.degree.basic
+
+/-!
+Given semirings `R` and `S` with a morphism `f : R →+* S`, we say that a polynomial with
+coefficients in `S` lifts if there exist `q : polynomial R` such that `map f p = q`.
+We prove various basic results about polynomials that lift, for example that if `p` and `q`
+lift then `p + q` lifts. Then, we prove that a polynomial that lifts can always be lifted to a
+polynomial of the same degree and that a monic polynomial that lifts can be lifted to a monic
+polynomial (of the same degree).
+
+## Main definition
+
+* `lifts R p` : the predicate that says that `p` lifts to `R`.
+
+## Main results
+
+* `lifts_deg_iff_lifts` : A polynomial lifts if and only if it can be lifted to a polynomial
+of the same degree.
+* `lifts_deg_monic_iff_lifts` : A monic polynomial lifts if and only if it can be lifted to a
+monic polynomial of the same degree.
+* `lifts_iff` : if `R` is commutative, a polynomial lifts if and only if it is in the image of
+`map_alg`, where `map_alg : polynomial R →ₐ[R] polynomial S` is the only `R`-algebra map
+that sends `X` to `X`.
+
+## Implementation details
+By definition, `p` lifts if and only if it is in the image of `map f`, that is the same
+as the image of `ring_hom.of (map f)`. We use the fact that the image is a semiring to prove all the
+basic results.
+
+Since we do not assume `R` to be commutative, we cannot say in general that the set of polynomials
+that lift is a subalgebra. (By `lift_iff` this is true if `R` is commutative.)
+
+-/
+
+open_locale classical big_operators
+noncomputable theory
+
+namespace polynomial
+universes u v w
+
+section semiring
+
+variables {R : Type u} [semiring R] {S : Type v} [semiring S] {f : R →+* S}
+
+/--We say that a polyonmial lifts when there is a polynomial `q` such that `map f q = p`. -/
+def lifts (f : R →+* S) (p : polynomial S) : Prop := ∃ (q : polynomial R), map f q = p
+
+lemma lifts_iff (p : polynomial S) : lifts f p ↔ p ∈ ring_hom.srange (ring_hom.of (map f)) :=
+  by simp only [lifts, ring_hom.coe_of, ring_hom.mem_srange]
+
+lemma lifts_iff_set_range (p : polynomial S) : lifts f p ↔ p ∈ set.range (map f) :=
+  by simp only [lifts, set.mem_range]
+
+/--The polynomial `0` lifts. -/
+lemma zero_lifts (f : R →+* S) : lifts f (0 : polynomial S) :=
+  by rw [lifts_iff]; exact subsemiring.zero_mem (ring_hom.of (map f)).srange
+
+/--The polynomial `1` lifts. -/
+lemma one_lifts (f : R →+* S) : lifts f (1 : polynomial S) :=
+  by rw [lifts_iff]; exact subsemiring.one_mem (ring_hom.of (map f)).srange
+
+/--If `(r : R)`, then `C (algebra_map R S r)` lifts. -/
+lemma lifts_of_C (f : R →+* S) (r : R) : lifts f (C (f r)) :=
```

Comment:
```suggestion
lemma lifts_C (f : R →+* S) (r : R) : lifts f (C (f r)) :=
```
Could you also provide a version `lifts_C'` that show `lifts f (C s)` assuming `s \in set.range f` (or a similar assumption).